src/SourceBuild/content/eng/tools/BinaryToolKit/BinaryTool.CompareBinariesAgainstBaselines.cs

```diff
@@ -0,0 +1,130 @@
+            }
+        }
+
+        Log.LogInformation("Finished comparing binaries.");
+
+        return binariesToRemove;
+    }
+
+    private IEnumerable<string> GetUnmatchedBinaries(
+        IEnumerable<string> searchFiles,
+        string? baselineFile,
+        string outputReportDirectory,
+        string targetDirectory,
+        Modes mode)
+    {
+        var patterns = ParseBaselineFile(baselineFile);
+
+        if (mode.HasFlag(Modes.Validate))
+        {
+            // If validating in any mode (Mode == Validate or Mode == All), 
+            // we need to detect both unused patterns and unmatched files.
+            // We simultaneously detect unused patterns and unmatched files for efficiency.
+
+            HashSet<string> unusedPatterns = new HashSet<string>(patterns);
+            HashSet<string> unmatchedFiles = new HashSet<string>(searchFiles);
+
+            foreach (string pattern in patterns)
+            {
+                Matcher matcher = new Matcher(StringComparison.Ordinal);
+                matcher.AddInclude(pattern);
+                
+                var matches = matcher.Match(targetDirectory, searchFiles);
+                if (matches.HasMatches)
+                {
+                    unusedPatterns.Remove(pattern);
+                    unmatchedFiles.ExceptWith(matches.Files.Select(file => file.Path));
+                }
+            }
+
+            UpdateBaselineFile(baselineFile, outputReportDirectory, unusedPatterns);
+
+            return unmatchedFiles;
+        }
+        else
+        {
+            // If only cleaning and not validating (Mode == Clean),
+            // we don't need to update the baseline files with unused patterns
+            // so we can just detect unmatched files.
+
+            Matcher matcher = new Matcher(StringComparison.Ordinal);
+            matcher.AddInclude("**/*");
+            matcher.AddExcludePatterns(patterns);
+
+            return matcher.Match(targetDirectory, searchFiles).Files.Select(file => file.Path);
+        }
+    }
+
+    private IEnumerable<string> ParseBaselineFile(string? file) {
+        if (!File.Exists(file))
+        {
+            return Enumerable.Empty<string>();
+        }
+
+        // Read the baseline file and parse the patterns, ignoring comments and empty lines
+        return File.ReadLines(file)
+            .Where(line => !string.IsNullOrWhiteSpace(line) && !line.StartsWith("#"))
+            .Select(line => line.Split('#')[0].Trim());
+    }
+
+    private void UpdateBaselineFile(string? file, string outputReportDirectory, HashSet<string> unusedPatterns)
+    {
+        if(File.Exists(file))
+        {
+            string updatedFile = Path.Combine(outputReportDirectory, "Updated" + Path.GetFileName(file));
+
+            File.Copy(file, updatedFile, true);
+            File.AppendAllText(updatedFile, $"{Environment.NewLine}{PatternsToRemoveMarker}{Environment.NewLine}");
```

Comment:
Why not just emit the file with all unused patterns removed? Else emit the unused entries to the logger and not emit an updated baseline?

Reply:
I felt like emitting the file with unused patterns might be counter-intuitive to developer productivity because there may be exclusions that are unused currently but expected to be used later. (i.e. if a binary is commonly removed and readded). I can change this to emit the unused entries to the logger.